temporalio/worker/_workflow.py

```diff
@@ -421,3 +439,97 @@ def nondeterminism_as_workflow_fail_for_types(self) -> Set[str]:
+
+    def __init__(self, message: str, replacement_tb: Optional[TracebackType] = None):
+        """Create a new DeadlockError, with message `msg` and optionally a traceback `tb` to be swapped in later.
+
+        Args:
+            message: Message to be presented through exception.
+            replacement_tb: Optional TracebackType to be swapped later.
+        """
+        super().__init__(message)
+        self._new_tb = replacement_tb
+
+    def has_pending_tb(self) -> bool:
+        """Check if the DeadlockError has a pending traceback. For use with swap_traceback().
+
+        Returns:
+            True if there is a pending traceback, false if not.
+        """
+        if self._new_tb:
+            return True
+        return False
+
+    def swap_traceback(self) -> None:
+        """Swap the current traceback for the replacement passed during construction. Used to work around Python adding the current frame to the stack trace.
+
+        Returns:
+            None
+        """
+        if self._new_tb:
+            self.__traceback__ = self._new_tb
+            self._new_tb = None
+
+    @classmethod
+    def from_deadlocked_workflow(
+        cls, workflow: WorkflowInstance, timeout: Optional[int]
+    ):
+        msg = f"[TMPRL1101] Potential deadlock detected: workflow didn't yield within {timeout} second(s)."
+        tid = workflow.get_thread_id()
+        if not tid:
+            return cls(msg)
+
+        tb, warn = cls._gen_tb_helper(tid)
+
+        if not tb:
+            msg += " Worker was unable to retrieve workflow thread info. No frames from workflow available."
+            return cls(msg)
+
+        if warn == "frame":
+            msg += " System encountered errors while collecting stack frames, minimal or partial traceback will be presented."
+        elif warn == "tb":
+            msg += " System encountered errors while constructing traceback, minimal or partial traceback will be presented."
+
+        return cls(msg, tb)
+
+    @staticmethod
+    def _gen_tb_helper(
+        tid: int,
+    ) -> Tuple[Optional[TracebackType], Optional[Literal["frame", "tb"]]]:
+        """Take a thread id and construct a stack trace.
+
+        Returns a tuple:
+            0 <Optional[TracebackType]> the traceback that was constructed, None if the thread could not be found.
+            1 <Optional['frame', 'tb']> a flag determining whether an additional warning should be appended to the error message, None if not needed.
+        """
+        frame = sys._current_frames().get(tid)
+        if not frame:
+            return (None, None)
+
+        # not using traceback.extract_stack() because it obfuscates the frame objects (specifically f_lasti)
+        warn: Optional[Literal["frame", "tb"]] = None
+        thread_frames = [frame]
+        while frame.f_back:
+            try:
+                frame = frame.f_back
+                thread_frames.append(frame)
+            except Exception:
```

Comment:
What are the exceptions that are raised here and below?

Reply:
I didn't think so either, but anything could change and we don't want it to fail into another error state when we're getting stack frames to present the original issue. (from a conversation with @cretz)
I'm going to be wrapping the entire original call in try/except (https://github.com/temporalio/sdk-python/pull/595#discussion_r1707095909) for the benefit of catch-all as I kind of agree the verbose errors are nice but not need-to-have and don't give the user any extra context they might need to debug a deadlock/timeout situation.